### PR TITLE
Feature 59 load images from storage

### PR DESCRIPTION
**Pull request by jurikmeh**

### CODE REVIEW

File: PhotoMap/Features/Timeline/ViewModels/TimelineViewModel.swift [hunk 1]
```diff
@@ -17,12 +17,14 @@ class TimelineViewModel: TimelineViewModelType {
     private var headerTitles = [String]()
     private var searchingMarkers = [String: [Marker]]()
     private var searchingHeaderTitles = [String]()
+    private let diContainer: DIContainerType
```
Comment (jurikmeh) [inline PhotoMap/Features/Timeline/ViewModels/TimelineViewModel.swift:20]: remove this line

File: PhotoMap/Features/Timeline/ViewModels/TimelineViewModel.swift [hunk 2]
```diff
@@ -17,12 +17,14 @@ class TimelineViewModel: TimelineViewModelType {
...
     private let cancelBag = CancelBag()
     
     // MARK: - Lifecycle
     init(coordinator: TimelineCoordinator, diContainer: DIContainerType) {
         self.coordinator = coordinator
         self.firestoreService = diContainer.resolve()
+        self.diContainer = diContainer
```
Comment (jurikmeh) [inline PhotoMap/Features/Timeline/ViewModels/TimelineViewModel.swift:27]: remove this line

File: PhotoMap/Core/Base/Models/Marker.swift [hunk 3]
```diff
@@ -13,18 +13,24 @@ struct Marker {
...
+    let images: [String]
+    let location: GeoPoint?
 }
 
 extension Marker {
+    var imageURL: URL? {
+        let documentDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
+        let fileName = "\(self.date.toString).png"
+        return documentDirectory?.appendingPathComponent(fileName)
+    }
```
Comment (atereshkov) [inline PhotoMap/Core/Base/Models/Marker.swift:25]: Can we actually test it in unit/integration tests?

File: PhotoMap/Features/Timeline/Controllers/MarkerCell.swift [hunk 4]
```diff
@@ -5,21 +5,48 @@
...
     @IBOutlet private weak var markerCategory: UILabel!
     @IBOutlet private weak var activityIndicator: UIActivityIndicatorView!
     
+    var viewModel: TimelineCellViewModel?
+    private let cancelBag = CancelBag()
+    
     func configure(with marker: Marker) {
-        markerImage.image = UIImage(systemName: "photo")
+        bind()
+        if let url = marker.imageURL, let imageData = try? Data(contentsOf: url), let image = UIImage(data: imageData) {
```
Comment (atereshkov) [inline PhotoMap/Features/Timeline/Controllers/MarkerCell.swift:23]: `marker.imageURL` is local file url or remote resource?

Comment (atereshkov) [reply PhotoMap/Features/Timeline/Controllers/MarkerCell.swift:23]: All the logic `if let` etc should not be in cell. Just try to bind data to UI

File: PhotoMap/Service/FirestoreService/FirestoreService.swift [hunk 5]
```diff
@@ -117,6 +117,37 @@ final class FirestoreService: FirestoreServiceType {
...
+    
+    func downloadImage(for marker: Marker) -> Future<URL?, FirestoreError> {
+        Future { [weak self] promise in
+            guard let currentUserId = self?.currentUserId else {
+                promise(.failure(.noCurrentUserId))
+                return
+            }
+            
+            guard let photoReference = self?.storage
+                    .child(currentUserId+Separator.slash+marker.date.toString+Separator.point+Path.imageType)
```
Comment (atereshkov) [inline PhotoMap/Service/FirestoreService/FirestoreService.swift:129]: This looks strange :)

File: PhotoMap/Service/FirestoreService/FirestoreService.swift [hunk 6]
```diff
@@ -117,6 +117,37 @@ final class FirestoreService: FirestoreServiceType {
...
+            else {
+                promise(.failure(.custom("can't find image reference in storage")))
```
Comment (atereshkov) [inline PhotoMap/Service/FirestoreService/FirestoreService.swift:131]: Should be at least localized

**Review by atereshkov — CHANGES_REQUESTED** [review]

File: PhotoMapTests/ViewModels/TimelineCellViewModelTests.swift [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        
+        XCTAssertNotNil(viewModel.description)
+        XCTAssertNotNil(viewModel.date)
+        XCTAssertNotNil(viewModel.category)
+        XCTAssertNotNil(viewModel.image)
+        XCTAssertTrue(firestoreService.downloadImageEndWithImage)
+        XCTAssertNotEqual(viewModel.image, UIImage(systemName: "photo"))
+    }
+    
+    func testIfHavePhotoLoadItFromCacheDocuments() {
```
Comment (atereshkov) [inline PhotoMapTests/ViewModels/TimelineCellViewModelTests.swift:43]:
```suggestion
    func test_IfImageExists_LoadItFromCacheDocuments() {
```

**Review by atereshkov — APPROVED** [review]

**Pull request merged**